### PR TITLE
🧪 #102-restart-tasks-v3

### PR DESCRIPTION
**Pull request by SiddiqueAhmad**
currently added restart connector with all or failed tasks

﻿

### CODE REVIEW

**Comment by SiddiqueAhmad** [comment]
there is one issue if we create a connector with main branch, so tests are simple currently

File: src/main/java/org/kcctl/command/RestartConnectorCommand.java [hunk 1]
```diff
@@ -22,25 +22,51 @@
...
 public class RestartConnectorCommand implements Runnable {
 
-    @Inject
-    ConfigurationContext context;
-
     @Parameters(paramLabel = "NAME", description = "Name of the connector (e.g. 'my-connector')", completionCandidates = ConnectorNameCompletions.class)
     String name;
 
+    @CommandLine.Option(names = { "-t", "--tasks" }, description = "Restart tasks either all or just failed")
+    String withTasks;
```
Comment (gunnarmorling) [inline src/main/java/org/kcctl/command/RestartConnectorCommand.java:36]: Let's make this an enum, as e.g. [here](https://github.com/kcctl/kcctl/blob/main/src/main/java/org/kcctl/command/PatchLogLevelCommand.java#L43), that way options will be validated and we get tab completion too.

**Review by gunnarmorling — COMMENTED** [review]
One comment inline; and could you update the completion script, too, please (see README for instructions)? Thx!

File: src/main/java/org/kcctl/service/KafkaConnectApi.java [hunk 2]
```diff
@@ -18,12 +18,7 @@
 import java.util.List;
 import java.util.Map;
 
-import javax.ws.rs.DELETE;
-import javax.ws.rs.GET;
-import javax.ws.rs.POST;
-import javax.ws.rs.PUT;
-import javax.ws.rs.Path;
-import javax.ws.rs.PathParam;
+import javax.ws.rs.*;
```
Comment (gunnarmorling) [inline src/main/java/org/kcctl/service/KafkaConnectApi.java:21]: As a best practice, let's avoid the star import, as it's prone to confusions with types from other packages.

Comment (SiddiqueAhmad) [reply src/main/java/org/kcctl/service/KafkaConnectApi.java:21]: true, intellij did it, but I will take care

**Comment by SiddiqueAhmad** [comment]
completion script give following error on this branch, but run fine on alpha5, 

`siddique@siddique-PC /media/siddique/WINDOWS101/open-source/kcctl (#102-restart-tasks-v3)$ java -cp "target/quarkus-app/app/*:target/quarkus-app/lib/main/*:target/quarkus-app/quarkus-run.jar"   picocli.AutoComplete -n kcctl --force org.kcctl.command.KcCtlCommand
picocli.CommandLine$InitializationException: Could not inject spec
	at picocli.CommandLine$Model$CommandReflection.initFromAnnotatedTypedMembers(CommandLine.java:11582)
	at picocli.CommandLine$Model$CommandReflection.initFromAnnotatedFields(CommandLine.java:11516)
	at picocli.CommandLine$Model$CommandReflection.extractCommandSpec(CommandLine.java:11449)
	at picocli.CommandLine$Model$CommandSpec.forAnnotatedObject(CommandLine.java:6229)
	at picocli.CommandLine.<init>(CommandLine.java:227)
	at picocli.CommandLine.toCommandLine(CommandLine.java:3526)
	at picocli.CommandLine.access$15600(CommandLine.java:145)
	at picocli.CommandLine$Model$CommandReflection.initSubcommands(CommandLine.java:11481)
	at picocli.CommandLine$Model$CommandReflection.extractCommandSpec(CommandLine.java:11447)
	at picocli.CommandLine$Model$CommandSpec.forAnnotatedObject(CommandLine.java:6229)
	at picocli.CommandLine.<init>(CommandLine.java:227)
	at picocli.CommandLine.<init>(CommandLine.java:221)
	at picocli.AutoComplete$App.call(AutoComplete.java:164)
	at picocli.AutoComplete$App.call(AutoComplete.java:96)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at picocli.AutoComplete.main(AutoComplete.java:74)
Caused by: picocli.CommandLine$PicocliException: Could not get scope for field picocli.CommandLine$Model$CommandSpec org.kcctl.command.InfoCommand.spec
	at picocli.CommandLine$Model$FieldBinding.set(CommandLine.java:11712)
	... 23 more
Caused by: picocli.CommandLine$InitializationException: Cannot instantiate org.kcctl.command.InfoCommand: the class has no constructor
	at picocli.CommandLine$DefaultFactory.create(CommandLine.java:5536)
	at picocli.CommandLine$Model$CommandUserObject.getInstance(CommandLine.java:11863)
	at picocli.CommandLine$Model$CommandUserObject.get(CommandLine.java:11888)
	at picocli.CommandLine$Model$FieldBinding.set(CommandLine.java:11711)
	... 23 more
`

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot, @SiddiqueAhmad. Some comments inline.

In regards to the completion problem, it seems like all the command types need to have default constructors (i.e. no parameters) for the completion generator to work. This is not the case any longer after the recent work towards Testcontainers-based tests. Could you therefore add these constructors, initializing all final fields to null. Try whether they can be private or at least package-private (i.e. no visibility modifier). And there should be a JavaDoc comment on them, saying "Only intended to be used during completion script generation".

Thanks again!

**Comment by gunnarmorling** [comment]
Hi, could you rebase this one to the latest upstream/main and force-push to this PR? It contains many unrelated commits right now.

**Comment by SiddiqueAhmad** [comment]
sure I will, additional constructor didn't solved the issue though, I will need your help

**Comment by gunnarmorling** [comment]
What's the exception you are seeing now? Note _all_ the command types will need such default constructor.

**Comment by gunnarmorling** [comment]
Going to close this one as it became stale and would at least need a rebase.

**Pull request closed without merging**